File: Advanced Robotics Course/Chapter2/Lesson2/Chapter2_Lesson2.java
import java.util.*;

class Edge {
    public final int to;
    public final double cost;
    public Edge(int to, double cost) {
        this.to = to;
        this.cost = cost;
    }
}

public class AStar {
    public static class NodeRecord implements Comparable<NodeRecord> {
        public final int v;
        public final double f;
        public final double g;
        public NodeRecord(int v, double f, double g) {
            this.v = v; this.f = f; this.g = g;
        }
        @Override
        public int compareTo(NodeRecord other) {
            return Double.compare(this.f, other.f);
        }
    }

    public static class Result {
        public final double cost;
        public final List<Integer> path;
        public Result(double cost, List<Integer> path) {
            this.cost = cost; this.path = path;
        }
    }

    public static Result astar(
        List<List<Edge>> graph,
        int start,
        int goal,
        java.util.function.DoubleUnaryOperator heuristic)
    {
        int n = graph.size();
        double[] g = new double[n];
        int[] parent = new int[n];
        boolean[] closed = new boolean[n];
        Arrays.fill(g, Double.POSITIVE_INFINITY);
        Arrays.fill(parent, -1);
        PriorityQueue<NodeRecord> open = new PriorityQueue<>();

        g[start] = 0.0;
        open.add(new NodeRecord(start, heuristic.applyAsDouble(start), 0.0));

        while (!open.isEmpty()) {
            NodeRecord rec = open.poll();
            int v = rec.v;
            if (closed[v]) continue;
            if (v == goal) {
                LinkedList<Integer> path = new LinkedList<>();
                for (int u = goal; u != -1; u = parent[u]) {
                    path.addFirst(u);
                }
                return new Result(rec.g, path);
            }
            closed[v] = true;
            for (Edge e : graph.get(v)) {
                if (e.cost < 0.0)
                    throw new IllegalArgumentException("Negative edge cost");
                double gTent = rec.g + e.cost;
                if (gTent < g[e.to]) {
                    g[e.to] = gTent;
                    parent[e.to] = v;
                    double f = gTent + heuristic.applyAsDouble(e.to);
                    open.add(new NodeRecord(e.to, f, gTent));
                }
            }
        }
        throw new RuntimeException("No path found");
    }
}
      

File: Tutorials/Codes/Advanced_Robotics_Course/Chapter2/Lesson2/Chapter2_Lesson2.m
function [path, cost] = astar_grid(occGrid, start, goal)
% occGrid: logical matrix, true for obstacle
% start, goal: [row, col]

[nr, nc] = size(occGrid);
INF = inf;

% 4-connected neighborhood
dirs = [ -1 0; 1 0; 0 -1; 0 1 ];

g = INF * ones(nr, nc);
parent = zeros(nr, nc, 2);

    function h = heuristic(cell)
        % Manhattan distance
        h = abs(cell(1) - goal(1)) + abs(cell(2) - goal(2));
    end

open = java.util.PriorityQueue();
startH = heuristic(start);
g(start(1), start(2)) = 0;
open.add({startH, start, 0}); % {f, [r c], g}

closed = false(nr, nc);

while ~open.isEmpty()
    entry = open.poll();
    cell = entry{2};
    r = cell(1); c = cell(2);
    g_val = entry{3};

    if closed(r, c)
        continue;
    end
    if r == goal(1) && c == goal(2)
        % reconstruct path
        path = [r, c];
        while parent(r, c, 1) ~= 0
            pr = parent(r, c, 1);
            pc = parent(r, c, 2);
            path = [pr, pc; path]; %#ok<AGROW>
            r = pr; c = pc;
        end
        cost = g_val;
        return;
    end

    closed(r, c) = true;
    for k = 1:size(dirs, 1)
        rr = r + dirs(k,1);
        cc = c + dirs(k,2);
        if rr < 1 || rr > nr || cc < 1 || cc > nc
            continue;
        end
        if occGrid(rr, cc)
            continue;
        end
        gTent = g_val + 1;
        if gTent < g(rr, cc)
            g(rr, cc) = gTent;
            parent(rr, cc, :) = [r, c];
            f = gTent + heuristic([rr, cc]);
            open.add({f, [rr, cc], gTent});
        end
    end
end

error('No path found');
end
      
